### PR TITLE
vmlatency, client: Add context to client layer

### PR DESCRIPTION
**Pull request by orelmisan**
Currently, functions from the client layer does not accept a context as a parameter, thus making them un-cancelable from the outside.

Add a context parameter to applicable functions from the client layer, in order to be able to cancel them from the outside.

This will be used in a follow-up PR to enforce the timeout parameter specified by the user.

### CODE REVIEW

**Review by RamLavi — COMMENTED** [review]
question, why is this important to us?
why not wait for https://github.com/kubevirt/kubevirt/issues/8885 to be approved?

**Comment by orelmisan** [comment]
> question, why is this important to us? why not wait for [kubevirt/kubevirt#8885](https://github.com/kubevirt/kubevirt/issues/8885) to be implemented?

Currently, the user-supplied `timeout` field it not respected.
In order to respect it, we need to pass a context to long running functions.

There is an open PR of the required changes to VMI methods in KubeVirt's client-go: https://github.com/kubevirt/kubevirt/pull/8908

The change proposed in this PR is a temporary workaround, until the clean fix will be merged and released so we can consume it.

**Comment by orelmisan** [comment]
Rebase

File: checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go [hunk 1]
```diff
@@ -79,9 +79,22 @@ func (c *Client) GetVirtualMachineInstance(ctx context.Context, namespace, name
...
+		createdVMI, err := c.KubevirtClient.VirtualMachineInstance(namespace).Create(vmi)
+		resultCh <- resultWrapper{createdVMI, err}
+	}()
+
+	select {
+	case result := <-resultCh:
+		return result.vmi, result.err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
```
Comment (ormergi) [inline checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go:97]: I see this logic reaping, how about reusing it by extracting it to function?
For example: 
```
func execWithContext(ctx context.Context, fn func() (any, error)) (any, error) {
	type result struct {
		value any
		err   error
	}

	resultCh := make(chan result, 1)

	go func() {
		v, err := fn()
		resultCh <- resultWrapper{v, err}
	}()

	select {
	case result := <-resultCh:
		return result.value, result.err
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}
```
Usage:
```
func (c *Client) CreateVirtualMachineInstance(ctx context.Context, namespace string, vmi *kvcorev1.VirtualMachineInstance) (*kvcorev1.VirtualMachineInstance, error) {
return execWithContext(ctx, func() (any, error) {
	return c.KubevirtClient.VirtualMachineInstance(namespace).Create(vmi)
})
```

Comment (orelmisan) [reply checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go:97]: This is a nice suggestion, thank you.
The parameter types of `Get()` and `Create()` are different.
Do you have a suggestion on how to bypass this problem?

Comment (EdDev) [reply checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go:97]: If I am not mistaken (please correct me here if I am), the usage of `any` is not for generics, but for pure interfaces. A solution that will not be type safe and require conversion at the caller.

Comment (orelmisan) [reply checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go:97]: `any` is an alias for `interface{}`.

File: checkups/kubevirt-vm-latency/vmlatency/internal/vmi/operation.go [hunk 2]
```diff
@@ -35,16 +35,19 @@ import (
...
 
 type KubevirtVmisClient interface {
-	GetVirtualMachineInstance(namespace, name string) (*kvcorev1.VirtualMachineInstance, error)
-	CreateVirtualMachineInstance(namespace string, vmi *kvcorev1.VirtualMachineInstance) (*kvcorev1.VirtualMachineInstance, error)
-	DeleteVirtualMachineInstance(namespace, name string) error
+	GetVirtualMachineInstance(ctx context.Context, namespace, name string) (*kvcorev1.VirtualMachineInstance, error)
+	CreateVirtualMachineInstance(
+		ctx context.Context,
+		namespace string,
+		vmi *kvcorev1.VirtualMachineInstance) (*kvcorev1.VirtualMachineInstance, error)
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply checkups/kubevirt-vm-latency/vmlatency/internal/vmi/operation.go:42]: Thanks for the suggestion.
I personally prefer to have the parameters names in order to have more context.
This multi-line declaration is indeed ugly, but in the balance between pretty and less informative, and ugly and informative, I choose the later.

**Review by EdDev — APPROVED** [review]
Very nice, thank you!

File: checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go (hunk 1 above)
Comment (EdDev) [reply checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go:97]: Yes, and it was aimed for generics and not interfaces. Otherwise, it is hard to follow the intention.
Bottom line, use `any` when intending to use generics and `interface{}` when intending to use interfaces to reduce confusions.

Comment (ormergi) [reply checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go:97]: >  Do you have a suggestion on how to bypass this problem?

I need to think about it more, lets's continue it on a follow-up PR.
